pyquil/operator_estimation.py

```diff
@@ -479,3 +541,93 @@ def measure_observables(qc: QuantumComputer, tomo_experiment: TomographyExperime
                 expectation=np.asscalar(obs_mean),
                 stddev=np.asscalar(np.sqrt(obs_var)),
             )
+
+
+def _get_diagonalizing_basis(list_of_pauli_terms: List[PauliTerm]):
+    """
+    Find the Pauli Term with the most non-identity terms
+
+    :param list_of_pauli_terms: List of Pauli terms to check
+    :return: The highest weight Pauli Term
+    """
+    # create an unsorted (frozen)set of tuples of the form (qubit, operation)
+    unsorted_qubit_ops = reduce(lambda x, y: x | y, (pt.operations_as_set() for pt in list_of_pauli_terms))
+    # create a list of tuples of the form (qubit, operation), sorted by qubits
+    qubit_ops = sorted(list(unsorted_qubit_ops), key=lambda x: x[0])
+    # return a PauliTerm created from a list of tuples of the form (operation, qubit)
+    return PauliTerm.from_list([(i[1], i[0]) for i in qubit_ops])
+
+
+def _max_tpb_overlap(expt_setting: ExperimentSetting, diagonal_sets: Dict):
+    """
+    Calculate the max overlap of an ExperimentSetting with other ExperimentSettings specified
+    by keys of diagonal_sets.
+    Returns a different key if ``expt_setting`` isn't diagonal in the same tpb as any of the keys.
+    If it is diagonal in the same tpb as some key, then the corresponding value (a list of ExperimentSettings)
+    is updated to include ``expt_setting`` and the key is updated so it has the largest weight PauliTerm
+    specifying the tpb.
+
+    :param expt_setting: ExperimentSetting
+    :param diagonal_sets: dictionary keyed with ExperimentSetting, and with each value being a
+            list of ExperimentSettings
+    :return: the updated diagonal_sets dictionary updated with the input expt_setting placed
+            appropriately into it
+    """
+    # loop through dict items
+    for key, es_list in diagonal_sets.items():
+        # determine if key is diagonal in the same tpb as expt_setting
+        b_diagonal_tpb = _expt_settings_diagonal_in_tpb(key, expt_setting)
```

Comment:
better variable name than `key`?

Reply:
put this line in an `if` to avoid nebulously named `b_diagonal_tpb`?

Reply:
Updated.